service/collector/rate_limit.go

```diff
@@ -121,61 +121,25 @@ func NewRateLimit(config RateLimitConfig) (*RateLimit, error) {
 }
 
 func (u *RateLimit) Collect(ch chan<- prometheus.Metric) error {
-	// We need all CRs to gather all subscriptions below.
-	var crs []providerv1alpha1.AzureConfig
-	{
-		mark := ""
-		page := 0
-		for page == 0 || len(mark) > 0 {
-			opts := metav1.ListOptions{
-				Continue: mark,
-			}
-			list, err := u.g8sClient.ProviderV1alpha1().AzureConfigs(metav1.NamespaceAll).List(opts)
-			if err != nil {
-				return microerror.Mask(err)
-			}
-
-			crs = append(crs, list.Items...)
-
-			mark = list.Continue
-			page++
-		}
+	clientSets, err := credential.GetAzureClientSetsFromCredentialSecrets(u.k8sClient, u.environmentName)
+	if err != nil {
+		return microerror.Mask(err)
 	}
 
-	// We generate clients and group them by subscription.
-	// Subscription is defined by the Secret used to fetch the credentials.
-	clientConfigBySubscription := map[string]client.AzureClientSetConfig{}
-	{
-		for _, cr := range crs {
-			config, err := credential.GetAzureConfig(u.k8sClient, key.CredentialName(cr), key.CredentialNamespace(cr))
-			if err != nil {
-				return microerror.Mask(err)
-			}
-			clientConfigBySubscription[config.SubscriptionID] = *config
-		}
-
-		// The operator potentially uses a different set of credentials than
-		// tenant clusters, so we add the operator credentials as well.
-		config := &client.AzureClientSetConfig{
-			ClientID:       u.cpAzureClientSetConfig.ClientID,
-			ClientSecret:   u.cpAzureClientSetConfig.ClientSecret,
-			SubscriptionID: u.cpAzureClientSetConfig.SubscriptionID,
-			TenantID:       u.cpAzureClientSetConfig.TenantID,
-		}
-		clientConfigBySubscription[config.SubscriptionID] = *config
+	// The operator potentially uses a different set of credentials than
+	// tenant clusters, so we add the operator credentials as well.
+	operatorClientSet, err := client.NewAzureClientSet(u.cpAzureClientSetConfig)
+	if err != nil {
+		return microerror.Mask(err)
 	}
+	clientSets[&u.cpAzureClientSetConfig] = operatorClientSet
 
 	ctx := context.Background()
 
 	// We track RateLimit metrics for each client labeled by SubscriptionID and
 	// ClientID.
 	// That way we prevent duplicated metrics.
-	for _, clientConfig := range clientConfigBySubscription {
-		azureClients, err := client.NewAzureClientSet(clientConfig)
-		if err != nil {
-			return microerror.Mask(err)
-		}
-
+	for clientConfig, clientSet := range clientSets {
```

Comment:
Would it make sense to extract loop body into function and execute it concurrently in [sync.ErrGroup](https://godoc.org/golang.org/x/sync/errgroup)?

Reply:
I will merge this and iterate on this on the [new azure-collector repository](https://github.com/giantswarm/azure-collector), if that's ok with you.